exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java

```diff
@@ -73,81 +78,173 @@
+    this(newBuilder()
+        .setConfig(config)
+        .setClusterCoordinator(coordinator));
   }
 
-  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, boolean isDirect)
-    throws OutOfMemoryException {
-    this(config, coordinator, null, isDirect);
+  /**
+   * @deprecated Create a DrillClient using {@link DrillClient.Builder}.
+   */
+  @Deprecated
+  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, boolean isDirect) {
+    this(newBuilder()
+        .setConfig(config)
+        .setClusterCoordinator(coordinator)
+        .setDirectConnection(isDirect));
   }
 
-  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, BufferAllocator allocator)
-      throws OutOfMemoryException {
-    this(config, coordinator, allocator, false);
+  /**
+   * @deprecated Create a DrillClient using {@link DrillClient.Builder}.
+   */
+  @Deprecated
+  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, BufferAllocator allocator) {
+    this(newBuilder()
+        .setConfig(config)
+        .setClusterCoordinator(coordinator)
+        .setAllocator(allocator));
+  }
+
+  /**
+   * @deprecated Create a DrillClient using {@link DrillClient.Builder}.
+   */
+  @Deprecated
+  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, BufferAllocator allocator,
+                     boolean isDirect) {
+    this(newBuilder()
+        .setConfig(config)
+        .setClusterCoordinator(coordinator)
+        .setAllocator(allocator)
+        .setDirectConnection(isDirect));
   }
 
-  public DrillClient(DrillConfig config, ClusterCoordinator coordinator, BufferAllocator allocator, boolean isDirect) {
-    // if isDirect is true, the client will connect directly to the drillbit instead of
-    // going thru the zookeeper
-    this.isDirectConnection = isDirect;
-    this.ownsZkConnection = coordinator == null && !isDirect;
-    this.ownsAllocator = allocator == null;
-    this.allocator = ownsAllocator ? RootAllocatorFactory.newRoot(config) : allocator;
-    this.config = config;
-    this.clusterCoordinator = coordinator;
+  // used by DrillClient.Builder
+  private DrillClient(final Builder builder) {
+    this.config = builder.config != null ? builder.config : DrillConfig.create();
+
+    this.ownsAllocator = builder.allocator == null;
+    this.allocator = !ownsAllocator ? builder.allocator : RootAllocatorFactory.newRoot(config);
+
+    this.isDirectConnection = builder.isDirectConnection;
+    this.ownsZkConnection = builder.clusterCoordinator == null && !isDirectConnection;
+    this.clusterCoordinator = builder.clusterCoordinator; // could be null
+
+    this.ownsEventLoopGroup = builder.eventLoopGroup == null;
+    this.eventLoopGroup = !ownsEventLoopGroup ? builder.eventLoopGroup :
+        TransportCheck.createEventLoopGroup(config.getInt(ExecConstants.CLIENT_RPC_THREADS), "Client-");
+
+    this.ownsExecutor = builder.executor == null;
```

Comment:
what is the motivation behind moving initiating `eventLoopGroup` and `executor` from `connect(String connect, Properties props)` to here, assuming the `connect` method was never called more than once ?